### PR TITLE
[R3C-465] MCH: pedestals calibration workflow

### PR DESCRIPTION
**Pull request by aferrero2707**
This PR introduces the workflows needed to process pedestal data, build the list of bad/noisy channels and store the list in the CCDB. The bad/noisy channels list will be used by the DCS to mute the corresponding channels during front-end configuration, and by the reconstruction in order to know which channels are expected to have no signals.

The processing workflow, in a simplified form, looks like this:
```
o2-mch-pedestals-decoding-workflow -b --patch-rdh | o2-calibration-mch-channel-calib-workflow -b --pedestal-threshold 200 --noise-threshold 2.0 | o2-calibration-ccdb-populator-workflow -b
```

In reality the decoding workflows should run in parallel on various EPNs, while the calibration workflow should aggregate the information from the various decoders and compute the channel-by-channel mean and RMS of the pedestals.
Once the end-of-stream is reached the calibrator scans the pedestals and compares the values with the user-supplied thresholds. Channels exceeding either threshold are added to the bad/noisy list, which is finally inserted into the CCDB.

The CDDB data consists of a vector of 32-bit integers, each integer representing the unique ID of a readout channel.

### CODE REVIEW

File: Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h
Comment (shahor02) [inline Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h]: I think CI will require to redefine this to ``MCHChannelData() = default``

File: Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibratorAlt.h
Comment (shahor02) [inline Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibratorAlt.h:12]: this class seem not to be used/compiled, but it has the same guards as https://github.com/AliceO2Group/AliceO2/blob/445d4cff6ca5f0385beff377b7535916a8c41c42/Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h#L11-L12, should be changed

File: Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorAlt.cxx
Comment (shahor02) [inline Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorAlt.cxx:51]: if not needed, better remoce the commented code

File: Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h
Comment (shahor02) [inline Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h]: if your calibration object should contain nothing but the content of this vector, then it is indeed better to define ``SampaChannelId`` and ``BadChannelsVector`` in a single header in DataFormats, and use the latter as CCDB object.

**Review by shahor02 — COMMENTED** [review]
Hi @aferrero2707 
To me it looks fine, see a few minor comments below.

File: Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibratorAlt.h
Comment (aferrero2707) [reply Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibratorAlt.h:12]: This was a test of the alternative approach I proposed at the beginning, and the files were committed by mistake. I will remove them.

File: Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorAlt.cxx
Comment (aferrero2707) [reply Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorAlt.cxx:51]: See above

File: Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h
Comment (aferrero2707) [reply Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h]: Changed.

File: Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h
Comment (aferrero2707) [reply Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h]: Done.

**Comment by aferrero2707** [comment]
Please do not merge this PR yet when all CI jobs will be green. I am still working on the interface with the QC.

File: DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h
Comment (aphecetche) [inline DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h:64]: Why do we need a ChannelGroup class which is nothing more than a ChannelId vector ?

File: DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h
Comment (aphecetche) [inline DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h:31]: that might not have to be solved for this PR, but ... we actually have two ways to reference the channels : (solarId,dsId,channel) (raw/online data oriented) and (deId,dsId,channel) (digit/offline oriented) and both can be encoded to 32 bits. 

I was wondering it we could not decide to call those channelIds (== uint32_t) and use a couple of bits in those 32 to indicate which convention is used (solar or de) ?

File: DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h
Comment (aferrero2707) [reply DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h:64]: In order to transform the vector into a root object. If there is an easier and more direct way to do that, I'm open to suggestions.

File: DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h
Comment (aferrero2707) [reply DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h:31]: Indeed we could use the most significant bit to indicate the type of encoding, and two different constructors (`DsChannelId(uint16_t solarId, uint8_t dsId, uint8_t channel)` or `DsChannelId(uint16_t deId, uint16_t padId)`) for the two cases.

As a starting point, I will add getter functions for the solarId/dsId/channel values, to avoid the direct manipulation of the 32-bit encoded value.

**Comment by aferrero2707** [comment]
@TimoWilken @ktf the fullCI check fails although there seem not to be any build errors, only warnings... any idea?

**Comment by aferrero2707** [comment]
@aphecetche @shahor02 for me the code in this PR is complete and ready to be merged once the CI jobs turn al green... I know that parts of the code are still perfectible, but I would propose to get this out for MW3 and then further improve the code structure once the workflow is validated with the tests.

There is a QC PR that depends on this one, and which is also needed for MW3: https://github.com/AliceO2Group/QualityControl/pull/632

**Comment by TimoWilken** [comment]
@aferrero2707 (Cc: @ktf) that's a bug in the o2checkcode.sh build script in alidist. I've proposed a patch there to fix it, and I'll restart the test once the fix is merged.

**Comment by aferrero2707** [comment]
@TimoWilken thanks!

**Comment by aferrero2707** [comment]
@TimoWilken @ktf the o2-dataflow CI job completed again with a failure, although there are no errors in the log... if I understood correctly, this should have been fixed by alisw/alidist#2891, right?

**Comment by TimoWilken** [comment]
@aferrero2707 that PR only fixed the fullCI check; the o2-dataflow check has a different problem, which I'm currently in the process of fixing.

File: Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx
Comment (aphecetche) [inline Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx]: do I read correctly that this will break when running with mDebug=false as mTFcount will be zero then ?

Comment (aferrero2707) [reply Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx]: Good catch! I will put the mTFcout increment outside of the mDebug block.

**Comment by aferrero2707** [comment]
@TimoWilken the o2-dataflow job failed again, with no logs (`No logs found. Please check the aurora log.` message when looking for full build log).

**Comment by aferrero2707** [comment]
@aphecetche do you think we can approve this PR and merge once the CI jobs are all green?

**Comment by TimoWilken** [comment]
@aferrero2707 thanks for the report; it was missing a tag. I've fetched it and restarted the check.

**Review by aphecetche — APPROVED** [review]

**Pull request merged**
